wled00/FX.cpp

```diff
@@ -2585,6 +2584,111 @@ uint16_t WS2812FX::mode_bouncing_balls(void) {
 
   return FRAMETIME;
 }
+/*
+*  bouncing balls on a track track Effect modified from Air Cookie's bouncing balls
+*/
+// modified for balltrack mode
+typedef struct Ballt {
+  unsigned long lastBounceUpdate;
+  float mass; // could fix this to be = 1. if memory is an issue
+  float velocity;
+  float height;
+} ballt;
+
+uint16_t WS2812FX::ball_track(bool collide) {
+  //allocate segment data
+  uint16_t maxNumBalls = 16;
+  uint16_t dataSize = sizeof(ballt) * maxNumBalls;
+  if (!SEGENV.allocateData(dataSize)) return mode_static(); //allocation failed
+
+  Ballt* balls = reinterpret_cast<Ballt*>(SEGENV.data);
+
+  // number of balls based on intensity setting to max of 16 (cycles colors)
+  // non-chosen color is a random color
+  uint8_t numBalls = int(((SEGMENT.intensity * (maxNumBalls - 0.8f)) / 255) + 1);
+
+  unsigned long time = millis();
+
+  if (SEGENV.call == 0) {
+    for (uint8_t i = 0; i < maxNumBalls; i++) {
+      balls[i].lastBounceUpdate = time;
+      balls[i].velocity = 0;
+      while(abs(balls[i].velocity)<.5){ // at the start make sure they are all moving
+        balls[i].velocity=10*(-.5-float(random16(0, 10000)) / 10000.0); // time units are ms
```

Comment:
This shouldn't be in a loop really.
BTW I *think* what you want to have here is a velocity from -5 to 5 but what you are calculating is -15 to -5 which would match the loop condition anyways.

Reply:
If the loop you are referring to is the "while" loop it is present to make sure the initialized velocities are not too close to zero.
Your right about the error in - I will update this to -.5+float(...) that will make this as I intended. Great to have another set of eyes. Let me test this and update.

Reply:
Maybe have something like this to avoid the loop
``` c++
x = positive_value_in_range(); // x > 0,5 && x < 5 
if (random(0,1) < .5)
  x = -x;
```
The while loop with random can take infinite time, although it is not likely. But when running several times (once for each ball) it might sum up.
IMHO one should avoid it where possible.

I like the general idea though. I might implement it in the other project I am working on if you don't mind.

Reply:
Thanks for the suggestion - this code concept is cleaner. No problem if you want to implement these ideas elsewhere.